xla/service/gpu/cudnn_fused_mha_rewriter.cc

```diff
@@ -402,6 +405,102 @@ StatusOr<bool> IsSupportedBMM2(const HloInstruction* bmm_2,
+          bmm_1->dot_dimension_numbers().rhs_contracting_dimensions()));
+
+  std::vector<int64_t> seq_q =
+      GetDimensionVector(bmm_1->operand(0)->shape().dimensions(), seq_q_dims);
+
+  std::vector<int64_t> seq_k =
+      GetDimensionVector(bmm_1->operand(1)->shape().dimensions(), seq_k_dims);
+
+  // flash attention TODO: check contracting dim here
+  std::vector<int64_t> contracting_dims_bmm1 = GetDimensionVector(
+      bmm_1->operand(0)->shape().dimensions(),
+      bmm_1->dot_dimension_numbers().lhs_contracting_dimensions());
+  // for now, seq_q and seq_k should be equal for flash attention to work
+  // flash attention only supports fixed topology so we check if custom call is
+  // such topology by checking custom_call_name
+  TF_RET_CHECK(seq_q.size() == 1);
+  TF_RET_CHECK(seq_k.size() == 1);
+  auto is_fixed_topology = [=]() {
+    return is_causal_mask &&
+           (custom_call_name == kCudnnfMHAScaleBiasSoftmaxDropoutCallTarget ||
+            custom_call_name == kCudnnfMHAScaleBiasSoftmaxCallTarget ||
+            custom_call_name == kCudnnfMHAScaleMaskSoftmaxDropoutCallTarget ||
+            custom_call_name == kCudnnfMHAScaleMaskSoftmaxCallTarget);
+  };
+
+  // flash attention TODO: make sure fwd act (bmm_2->operand[0]) is only used by
+  // bmm2 and bwd bmm2_grad1
+  return (seq_q[0] == seq_k[0] && seq_q[0] > 512 && seq_q[0] % 128 == 0 &&
+          is_fixed_topology());
+}
+
+bool IsCausalMaskPattern(HloInstruction* mask) {
+  auto causal_mask_pattern_fwd_remat = m::Broadcast(OptionalBitcast(
+      m::Select(m::Compare(m::Iota(), m::Iota()), m::Broadcast(m::Constant()),
+                m::Broadcast(m::Constant()))));
+  auto causal_mask_pattern_bwd =
+      m::Broadcast(m::Convert(OptionalBitcast(m::Minimum(
+          m::Op(),
+          m::Broadcast(OptionalBitcast(m::Select(
+              m::Compare(m::Iota(), m::Iota()), m::Broadcast(m::Constant()),
+              m::Broadcast(m::Constant()))))))));
+  HloInstruction* param = nullptr;
+  HloInstruction* gte = nullptr;
+  auto causal_mask_pattern_fwd =
+      m::Broadcast(m::Bitcast(m::GetTupleElement(&gte, m::Parameter(&param))));
+  auto causal_mask_pattern = m::AnyOf<HloInstruction>(
+      causal_mask_pattern_fwd_remat, causal_mask_pattern_fwd,
+      causal_mask_pattern_bwd);
+  if (Match(mask, causal_mask_pattern)) {
+    if (param != nullptr) {
+      // need to track to outside of the while loop to find the real mask
+      // while body computation
+      auto mask_index = gte->tuple_index();
+      auto comp = param->parent();
+      auto mod = comp->parent();
+      auto name = comp->name();
+      auto entry_computation = mod->entry_computation();
+      bool is_causal_mask = true;
+      for (HloInstruction* instr :
+           entry_computation->MakeInstructionPostOrder()) {
+        if (instr->opcode() == HloOpcode::kWhile &&
+            instr->while_body()->name() == name) {
+          DCHECK(instr->operand(0)->opcode() == HloOpcode::kTuple);
```

Comment:
Prefer referring bad Status with TF_RET_CHECK, or CHECK if the error can absolutely not occur.